tools/oeedger8r/Emitter.ml

```diff
@@ -73,62 +71,49 @@ let conv_array_to_ptr (pd : pdecl) : pdecl =
         (PTPtr (tmp_aty, tmp_pa), tmp_declr)
       else (pt, declr)
 
-(** Note that, for a foreign array type [foo_array_t] we will generate
-    [foo_array_t* ms_field;] in the marshalling data structure to keep
-    the pass-by-address scheme as in the C programming language. *)
-let mk_ms_member_decl (pt : parameter_type) (declr : declarator)
-    (isecall : bool) =
-  (* TODO: Clean this up. *)
-  let aty = get_param_atype pt in
-  let tystr =
-    if is_foreign_array pt then
-      sprintf "/* foreign array of type %s */ void" (get_tystr aty)
-    else get_tystr aty
-  in
-  let ptr = if is_foreign_array pt then "*" else "" in
-  let field = declr.identifier in
-  (* String attribute is available for in/in-out both ecall and ocall.
-     For ocall ,strlen is called in trusted proxy code, so no need to
-     defense it. *)
-  let need_str_len_var (pt : parameter_type) =
-    match pt with
-    | PTVal _ -> false
-    | PTPtr (_, pa) ->
-        if pa.pa_isstr || pa.pa_iswstr then
-          match pa.pa_direction with
-          | PtrInOut | PtrIn -> if isecall then true else false
-          | _ -> false
-        else false
-  in
-  let str_len =
-    if need_str_len_var pt then sprintf "    size_t %s_len;\n" field else ""
+(** ----- End code borrowed and tweaked from {!CodeGen.ml} ----- *)
+
+(* Borrowed from upstream OCaml implementation in
+   https://github.com/ocaml/ocaml/pull/2185 *)
+let filter_map f =
+  let rec aux accu = function
+    | [] -> List.rev accu
+    | x :: l -> (
+      match f x with None -> aux accu l | Some v -> aux (v :: accu) l )
   in
-  let dmstr = get_array_dims declr.array_dims in
-  sprintf "    %s%s %s%s;\n%s" tystr ptr field dmstr str_len
+  aux []
 
-(** ----- End code borrowed and tweaked from {!CodeGen.ml} ----- *)
+(* Helper to flatten and map at the same time. *)
+let flatten_map f l =
+  let l = List.map f l in
+  List.flatten l
 
-let is_in_ptr (ptype, _) =
+let is_in_ptr (ptype : parameter_type) =
   match ptype with
   | PTVal _ -> false
   | PTPtr (_, a) -> a.pa_chkptr && a.pa_direction = PtrIn
 
-let is_out_ptr (ptype, _) =
+let is_out_ptr (ptype : parameter_type) =
   match ptype with
   | PTVal _ -> false
   | PTPtr (_, a) -> a.pa_chkptr && a.pa_direction = PtrOut
 
-let is_inout_ptr (ptype, _) =
+let is_inout_ptr (ptype : parameter_type) =
   match ptype with
   | PTVal _ -> false
   | PTPtr (_, a) -> a.pa_chkptr && a.pa_direction = PtrInOut
 
-let is_str_ptr (ptype, _) =
+let is_str_ptr (ptype : parameter_type) =
   match ptype with PTVal _ -> false | PTPtr (_, a) -> a.pa_isstr
 
-let is_wstr_ptr (ptype, _) =
+let is_wstr_ptr (ptype : parameter_type) =
   match ptype with PTVal _ -> false | PTPtr (_, a) -> a.pa_iswstr
 
+let gen_c_for i =
+  if i = "1" then [] else [sprintf "for (size_t _i = 0; _i < %s; ++_i)" i]
```

Comment:
I'd like to do something better than `i = "1"` here, but honestly, it's a true statement as `count = 1` is the assumed default unless `count` is explicitly set higher.

Also should probably use a less likely to collide variable name than `_i`.